Feat: 예약 날짜 및 시간, 인원 선택 컴포넌트 구현

itscold96 opened the pull request:
## ✏️ 작업 내용 요약
> 예약 날짜 및 시간, 인원 선택 컴포넌트 구현

## 📝 작업 내용 설명
- 캘린더에서 년, 월, 일 이동이 가능합니다.
- 오늘의 날짜는 테두리로 알아볼 수 있습니다.
- 예약 가능한 날짜만 캘린더에서 선택이 가능하며, 나머지 날짜는 클릭이 불가능합니다.
- 예약 가능 일자를 API 요청으로 받아오고 오늘 이전의 예약 가능 일자는 필터링하여 캘린더에 보이지 않습니다.
- 예약 가능 시간은 사용자의 로컬 시간을 기준으로 해당 시간이 예약 가능 체험의 시작 시간 이전일 경우의 시간들만 표출됩니다.
- 예약 가능 시간이 많아지더라도 레이아웃이 깨지지 않도록 수평 스크롤이 가능하도록 하였으며,
스크롤 기능은 하되 스크롤은 눈에 보이지 않도록 처리하였습니다.(모바일 환경에서는 터치로 스크롤 가능 확인)
- 사용자들의 예약 중복을 최대한 막기 위해, 연-월-일 이동 시 매번 서버에서 새로운 예약 가능 일정을 받아옵니다.
- 피그마 디자인 요구사항에서 PC & Tablet과 Moblie 레이아웃이 완전히 달라 두 개의 컴포넌트로 분리하였으며,
  반응형으로 두 컴포넌트가 전환되더라도 사용자가 입력한 데이터는 전역 상태로 처리하여 일관성을 유지하도록 하였습니다.
- useMediaQuery로 반응형 크기를 계산하고 있기 때문에, SSR 시 Mobile, Tablet 화면에서도 PC에서 보이는 컴포넌트가
  잠깐 보였다가 사라지는 현상을 발견하여 dynamic import 로 SSR시에 렌더링을 막아두었습니다.
  (예약 컴포넌트는 client component입니다.)
- 예약하기 API 결과에 따라 모달창을 띄워줍니다. 예약 실패의 경우에도 메시지를 모달로 알려줍니다.


## 📷 스크린샷

https://github.com/user-attachments/assets/68744d18-7e57-4b32-8c52-3977570974bf

## 🤓 공용 로직
- getCurrencyFormat(): 숫자를 인자로 받아 3자리 수마다 쉼표( , )를 찍어주는 문자열로 반환하는 함수입니다.
  - 여러 군데에서 사용될 수 있을 것 같아 util 함수로 제작해놓았습니다.
```tsx
const price = 1000;

getCurrencyFormat(price) // '1,000' 

// 아래와 같이 활용 가능합니다.
const perPersonPrice = `₩ ${getCurrencyFormat(price)}`; // '₩ 1,000'
```

- ResponsiveReservation: PC, 모바일 환경에서 서로 다른 레이아웃을 가진 컴포넌트 렌더링 해주는 컴포넌트
  - 윤하님께서 상세페이지에 예약 캘린더 사용하실 때, 따로 반응형 구현 없이 이 컴포넌트만 가져다 쓰시면 됩니다.
```tsx
// api 에서 받은 activityId와 price를 넣어주세요
<ResponsiveReservation activityId={2970} price={1000} />
``` 

- ReservationCalendar: 피그마 디자인 + 선택 가능한 날짜 필터링 기능이 있는 캘린더 컴포넌트
  - 성진님께서 예약 페이지에서 사용하실 수 있도록 예약 캘린더를 모듈화하였습니다.
  - Mantine Calendar의 다른 모든 프로퍼티들도 받을 수 있도록 처리해두었습니다.
  - 더 필요한 기능이 있으시면 사용부에서 직접 프로퍼티를 작성하여 사용이 가능합니다.
```tsx
<ReservationCalendar
  selectedDate={selectedDate} // 선택된 날짜
  onClickDate={setSelectedDate} // 날짜 클릭 시 실행될 함수, 자동으로 매개변수로 클릭한 날짜 (date:Date)를 받도록 제작하였습니다.
  availableDates={availableDates} // 선택 가능한 날짜를 (YYYY-MM-DD) 포맷 배열로 넣어주면 됩니다.
  // 필요한 것이 있다면 Mantine 공식 문서를 참조하여 아래와 같이 추가 작성하면 됩니다.
  onNextMonth={date => setSelectedDate(date)}
  onPreviousMonth={date => setSelectedDate(date)}
  highlightToday
/>
```

## 💬 리뷰 요구사항
> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
- 수정이나 개선이 필요한 부분이 있다면 알려주세요.

## 🏷️ 연관된 이슈 번호
> [#30 예약 캘린더 제작](#30)

## ✅ 체크리스트:
- [x] 내 코드는 이 프로젝트의 컨벤션을 따릅니다.
- [x] 나는 PR 전에 내 코드를 검토했습니다.
- [x] 내 코드에 이해하기 어려운 부분에는 별도로 주석을 추가했습니다.
- [x] 내 변경 사항으로 인해 다른 요소들에 새로운 에러가 발생하지 않습니다.

itscold96 committed:
#30 env: mantine ui provider 모듈화 및 테마 설정

itscold96 committed:
#30 feat: 통화 단위 변환 함수 모듈화

itscold96 committed:
#30 design: 예약 카드 퍼블리싱

itscold96 committed:
#30 feat: 선택 가능한 날짜만 선택할 수 있는 기능 구현 및 스타일 부여

itscold96 committed:
#30 feature: 예약 가능 시간 데이터 패칭 및 캘린더 연동

itscold96 committed:
#30 feat: 참여 인원수 조정 스탭퍼 구현

itscold96 committed:
#30 feat: 인원 수에 따라 총합 금액 변경 구현

itscold96 committed:
Merge branch 'dev' of https://github.com/codeit-final-project-team3/GiveMeFive into #30/feature/reservation

itscold96 committed:
#30 refactor: Reservation 컴포넌트 내부 모듈화

캘린더 영역, 인원수 증감 스탭퍼 모듈화

itscold96 committed:
#30 feat: 예약 가능 시간이 많아질 경우 스크롤 기능 추가

좌우 스크롤, 터치는 되지만 스크롤은 보이지 않도록 처리

itscold96 committed:
#30 feat: tablet 화면에서 날짜, 시간 선택 화면 모달로 분기 처리

itscold96 committed:
#30 refactor: AvailableTimeList 모듈화

itscold96 committed:
#30 feat: 태블릿 화면에서 선택된 날짜와 시간이 표기되는 기능 추가

itscold96 committed:
#30 refactor: 모바일 화면에서도 같은 로직을 활용하기 위해 useReservation 커스텀 훅 제작

itscold96 committed:
#30 feat: 예약 캘린더 모바일 화면에서의 분기 처리

itscold96 committed:
#30 feat: 모바일 화면 예약 캘린더 구현

itscold96 committed:
#30: refactor: 예약 관련 전역 상태 관리 훅 제작

itscold96 committed:
#30 refactor: 기존 커스텀 훅 및 props drilling 이었던 코드를 전역 상태 훅으로 리팩토링

itscold96 committed:
#30 feat: 선택된 날짜에 예약 가능한 시간이 없을 경우 처리

itscold96 committed:
#30 refactor: 예약 모달 모듈화

itscold96 committed:
#30 refactor: 캘린더 데이터 표출 및 setSelectedDate 수정

- 월 바꿀 때도 데이터 불러올 수 있게 처리
- 캘린더 리렌더링 되어도 선택된 날짜 그대로 보여주도록 처리
- setSelectedDate는 날짜만 받고, 예약 가능 날짜 처리는 useAvailableSchedule 쿼리 훅에서 하도록 수정
- 월 형식이 MM 형식으로 변환 과정 추가

itscold96 committed:
#30 feat: 예약 가능한 시간 배열, 현재 시점 이후로 필터링

itscold96 committed:
#30 refactor: 사용부에서 쉽게 쓸 수 있도록 화면에 따른 분기 컴포넌트 모듈화

itscold96 committed:
#30 feat: 예약 제출 및 결과에 따른 모달 처리 구현

itscold96 committed:
#30 fix: 헤더, 모달 쌓임 맥락 조정

itscold96 committed:
#30 refactor: 반응형 컴포넌트 변환 시, SSR로 인해 컴포넌트가 바뀌는 것이 보이는 현상 제거

itscold96 committed:
#30 fix: 로그인 하지 않고, 예약 진행 시 unahorized 가 아닌 한글 안내 메시지가 표출되도록 수정

KimDasom521 commented on src/utils/getCurrencyFormat.ts:
오오 예약 내역 페이지와 내 체험 관리 페이지에도 가격을 써야하는데 유용할 것 같습니다!

itscold96 committed:
#30 refactor: 예약 관리 페이지에서도 사용할 수 있도록 Calendar 모듈화

MELATONIN99 commented on src/app/components/reservation/Reservation.module.scss:
```diff
@@ -0,0 +1,53 @@
+  border: 1px solid $gray300;
+  border-radius: 12px;
+
+  box-shadow: 0px 4px 16px 0px #1122110d;
```
0px인 경우 px 을 생략하여 사용하는 걸 추천한다고 멘토링 때 들은 적이 있습니다.
0 자체는 이미 어떤 단위를 붙여도 값의 의미가 변하지 않고
당장에는 크게 영향을 끼치지 않지만 습관처럼 CSS를 다룰 때 불필요한 단위를 생략하면
전체 파일 크기를 줄일 수 있어 페이지 로딩 속도 개선에 도움이 될 것 같습니다!
(최적화 플러그인 마다 다르지만 webpack에서 번들링 될 때, 0에 대한 px값이 떼어지는 작업이 있다고 합니다)

itscold96 replied:
제가 직접 작성할 때는 0px면 0으로 씁니다만,
피그마에서 긁어서 쓰면 0px가 그대로 나와서 따로 신경을 안썼는데,
번들링에서 영향이 있는 건 처음 알았네요!
수정해놓겠습니다!
좋은 내용 배워갑니다ㅎㅎ

MELATONIN99 reviewed (approved):
전체적으로 코드의 분리가 잘 되어있고 주석의 설명이 잘 되어있어
많은 고민을 하신 게 느껴집니다! 수고하셨습니다

itscold96 committed:
#30 fix: useReservationMutation에서 쿼리 클라이언 사용 부분 수정

- const queryClient = new QueryClient();
  - 전역에서 Provider를 만들 때 사용.
- const queryClient = useQueryClient();
  - 이미 만들어진 전역 QueryClient를 가져 옴.

itscold96 committed:
#30 fix: 0px -> 0 으로 스타일 구문 수정

ayoooyh commented on src/app/components/providers/MantineProviders.tsx:
```diff
@@ -0,0 +1,17 @@
+export default function MantineProviders({ children }: { children: ReactNode }) {
+  return (
+    <MantineProvider>
+      {/* 새로 고침 시 백그라운드가 까맣게 변하는 현상 방지 */}
```
사용자 입장에서 놀랄 수 있는 부분을 잘 커버해주신 것 같습니다!!👍

itscold96 committed:
Merge branch 'dev' into #30/feature/reservation
